docs/ip_reuse.md

```diff
@@ -0,0 +1,53 @@
+# IP reuse
+
+## What is IP reuse
+
+As a developer or user there could be a need to make IP addresses or IP allocations that will be attached to the nodes in the chain of: **BMH <=> M3M <=> CAPI machine <=> node** objects predictable, for example during the upgrade process and the main goal of this feature is to make that possible.
+
+## Logic behind
+
+As is now, M3IPPool is an object representing a set of IP address pools to be used for IP address allocations. An M3IPClaim is an object representing a request for an IP address allocation. Consequently, the M3IPClaim name is structured as following:
+
+**M3IPClaimName** = **M3DataName** + **(-)** + **M3IPPoolName**
+
+Example: m3datatemplate-0-pool0
+
+First part in `M3IPClaimName`, `M3DataName` is always created in a format of `M3DataTemplateName-index` and second part is `M3IPPoolName` which is the name of Metal3IPPool object defined in the [IP Address manager repo](https://github.com/metal3-io/ip-address-manager).
+In the CAPM3 workflow, when Metal3Machine is created and M3Data is requested, the
+process of choosing an `index` to be appended in the name of `M3DataTemplateName` is random.
+For example, let's imagine we have two Metal3Machines: `metal3machine-0` and `metal3machine-1`
+which creates following `m3datatemplate-0` and `m3datatemplate-1` Metal3Data objects respectively. However, if two nodes are being upgraded at a time, there is no guarantee that same indexes will be appended to respective objects and in fact it can be in completely reverse order (i.e. `metal3machine-0` gets `m3datatemplate-1`).
+In order to make it predictable, we structure M3IPClaim name using the BareMetalHost name, as following:
```

Comment:
```suggestion
In the CAPM3 workflow, when a Metal3Machine is created and a M3Data is requested, the
process of choosing an `index` to be appended to the name of the `M3DataTemplateName`, is random.
For example, let's imagine we have two Metal3Machines: `metal3machine-0` and `metal3machine-1`
which creates the following `m3datatemplate-0` and `m3datatemplate-1` Metal3Data objects respectively. However, if two nodes are being upgraded at a time, there is no guarantee that same indices will be appended to the respective objects and in fact it can be in completely reverse order (i.e. `metal3machine-0` gets `m3datatemplate-1`).
In order to make it predictable, we structure M3IPClaim name using the BareMetalHost name, as following:
```